test/Database/LSMTree/ModelIO/Monoidal.hs

```diff
@@ -0,0 +1,315 @@
+     , Typeable k
+     , Typeable v
+     )
+  => SnapshotName
+  -> TableHandle m k v
+  -> m ()
+snapshot n TableHandle {..} = atomically $
+    withModel "snapshot" thSession thRef $ \tbl ->
+        modifyTVar' (snapshots thSession) (Map.insert n (toDyn tbl))
+
+-- | Open a table through a snapshot, returning a new table handle.
+open ::
+     ( IOLike m
+     , SomeSerialisationConstraint k
+     , SomeSerialisationConstraint v
+     , Typeable k
+     , Typeable v
+     )
+  => Session m
+  -> SnapshotName
+  -> m (TableHandle m k v)
+open s n = atomically $ do
+    ss <- readTVar (snapshots s)
+    case Map.lookup n ss of
+        Nothing -> throwSTM IOError
+            { ioe_handle      = Nothing
+            , ioe_type        = NoSuchThing
+            , ioe_location    = "open"
+            , ioe_description = "no such snapshot"
+            , ioe_errno       = Nothing
+            , ioe_filename    = Nothing
+            }
+
+        Just dyn -> case fromDynamic dyn of
+            Nothing -> throwSTM IOError
+                { ioe_handle      = Nothing
+                , ioe_type        = InappropriateType
+                , ioe_location    = "open"
+                , ioe_description = "table type mismatch"
+                , ioe_errno       = Nothing
+                , ioe_filename    = Nothing
+                }
+
+            Just tbl' -> do
+                ref <- newTMVar tbl'
+                i <- new_handle s ref
+                return TableHandle { thRef = ref, thId = i, thSession = s }
+
+{-------------------------------------------------------------------------------
+  Mutiple writable table handles
+-------------------------------------------------------------------------------}
+
+-- | Create a cheap, independent duplicate of a table handle. This returns a new
+-- table handle.
+duplicate ::
+     IOLike m
+  => TableHandle m k v
+  -> m (TableHandle m k v)
+duplicate TableHandle {..} = atomically $
+    withModel "duplicate" thSession thRef $ \tbl -> do
+        thRef' <- newTMVar tbl
+        i <- new_handle thSession thRef'
+        return TableHandle { thRef = thRef', thId = i, thSession = thSession }
+
+{-------------------------------------------------------------------------------
+  Merging tables
+-------------------------------------------------------------------------------}
+
+-- | Merge full tables, creating a new table handle.
+mergeTables ::
+     (IOLike m, SomeSerialisationConstraint v, SomeUpdateConstraint v)
+  => TableHandle m k v
+  -> TableHandle m k v
+  -> m (TableHandle m k v)
+mergeTables hdl1 hdl2
+{-
+    -- cannot == io-sim TVars.
+    | thSession hdl1 /= thSession hdl2
+    = throwSTM IOError
+        { ioe_handle      = Nothing
+        , ioe_type        = InappropriateType
+        , ioe_location    = "mergeTables"
+        , ioe_description = "different sessions"
+        , ioe_errno       = Nothing
+        , ioe_filename    = Nothing
+        }
+-}
```

Comment:
Is this something we have to fix?

Reply:
I don't know, I don't write the spec. (The spec doesn't say about this case anything, so it's UB).

Reply:
... and it would be helpful if there were `forall b. Eq (TVar m b)` instance, but that cannot be said as `TVar` is type-family.

Reply:
I say we let it be as is for now